File: problems/930_binary_subarrays_with_sum.py
class Solution:
    # Complexity: O(n) time, O(n) space.
    def numSubarraysWithSum(self, nums: list[int], goal: int) -> int:
        freq: dict[int, int] = {}
        sum = 0
        count = 0
        for num in nums:
            sum += num
            diff = sum - goal
            if diff == 0:
                count += 1
            if diff in freq:
                count += freq.get(diff, 0)

            freq[sum] = freq.get(sum, 0) + 1

        return count

File: problems/930-binary-subarrays-with-sum.ts
function numSubarraysWithSum(nums: number[], goal: number): number {
  const freq = new Map<number, number>()
  let sum = 0
  let count = 0

  for (const num of nums) {
    sum += num
    const diff = sum - goal
    if (diff === 0) {
      count++
    }
    if (freq.has(diff)) {
      count += freq.get(diff)!
    }
    freq.set(sum, (freq.get(sum) ?? 0) + 1)
  }

  return count
}
